moshi/src/main/java/com/squareup/moshi/JsonUtf8Writer.java

```diff
@@ -273,16 +276,34 @@ private void writeDeferredName() throws IOException {
     return this;
   }
 
-  @Override public JsonWriter value(BufferedSource source) throws IOException {
+  @Override public BufferedSink valueSink() throws IOException {
     if (promoteValueToName) {
       throw new IllegalStateException(
-          "BufferedSource cannot be used as a map key in JSON at path " + getPath());
+          "BufferedSink cannot be used as a map key in JSON at path " + getPath());
     }
     writeDeferredName();
     beforeValue();
-    sink.writeAll(source);
-    pathIndices[stackSize - 1]++;
-    return this;
+    pushScope(STREAMING_VALUE);
+    return Okio.buffer(new Sink() {
+      @Override public void write(Buffer source, long byteCount) throws IOException {
+        sink.write(source, byteCount);
+      }
+
+      @Override public void close() {
+        if (peekScope() != STREAMING_VALUE) {
+          throw new AssertionError();
+        }
+        stackSize--; // Remove STREAMING_VALUE from the stack.
+        pathIndices[stackSize - 1]++;
+      }
+
+      @Override public void flush() {
+      }
+
+      @Override public Timeout timeout() {
+        return Timeout.NONE;
```

Comment:
should this be the sink’s timeout?

Reply:
I thought it weird that someone streaming a value could potentially change the timeout or deadline of the enclosing JSON write.

Reply:
Neat